daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala

```diff
@@ -431,11 +437,27 @@ private[lf] final class Compiler(
         SBFromAny(ty)(compile(e))
       case ETypeRep(typ) =>
         SEValue(STypeRep(typ))
-      case EThrow(_, _, _) | EFromAnyException(_, _) | EToAnyException(_, _) =>
-        // TODO https://github.com/digital-asset/daml/issues/8020
-        sys.error("exceptions not supported")
+      case EToAnyException(ty, e) =>
+        val messageFunction = compileExceptionType(ty)
+        SBToAnyException(ty, messageFunction)(compile(e))
+      case EFromAnyException(ty, e) =>
+        SBFromAnyException(ty)(compile(e))
+      case EThrow(_, ty, e) =>
+        val messageFunction = compileExceptionType(ty)
+        SBThrow(SBToAnyException(ty, messageFunction)(compile(e)))
```

Comment:
1. Can you take me through the decision to wrap up the exception in an `AnyException` value here, before throwing? This was definitely part of an earlier design of the LF (but as we discussed, Martin & I decided to restrict throws in LF to only known exception types).

2. The message function is associated with the thrown value? Isn't this something you can get from the type `ty`?

Reply:
For 1, you need the type of the exception that is thrown so you can check it in the handler. I agree the current implementation is a bit confusing by reusing `AnyException` for this (and is perhaps missing unwrapping it in some places although I think I didn’t find any) but you need to tuple the type up with the exception somehow.

2, yeah, see my question on merging it with `SAny`.

Reply:
Ah, I agree you need to know the type when throwing it, just not sure it made sense to do it here instead of doing it in SBThrow (like the LF spec says).

Reply:
Oh I see, yeah agreed `SBThrow` is probably more sensible.

Reply:
There are two things a runtime value representing a value of type `AnyException` must be able to support: (1) `EFromAnyException(expect-ty)` and (2) `BAnyExceptionMessage`, so it seems natural for the runtime representation `SAnyException` to contain the triple of (payload-value,actual-ty,message-func)

It's possible, as Moritz suggests, that the message-func might be elided. However, this strikes me as slightly odd, as it requires the (new) speedy compiler function `compileExceptionType` to be called at speedy runtime.

Since both `BAnyExceptionMessage` and `EFromAnyException` can be applied to arbitrary expressions, the `SAnyException` must be created independently of throwing.

Regarding how `EThrow` is supported... the code in (speedy) Compiler make this clear:

```
      case EFromAnyException(ty, e) =>
        SBFromAnyException(ty)(compile(e))
      case EThrow(_, ty, e) =>
        val messageFunction = compileExceptionType(ty)
        SBThrow(SBToAnyException(ty, messageFunction)(compile(e)))
```

The LF expression `EThrow` is compiled so the two stages are made explicit: `SBToAnyException` and `SBThrow`

Although the LF semantics restricts throw to only known exception types, the speedy runtime version of throw (`SBThrow`) doesn't care. It can throw any value (which is excepted to have runtime representation `SAnyException`). This makes sense also when you consider the execution of selective handlers... if a handler refuses to handle (returns `None`) then the `AnyException` value being thrown has already been constructed and the re-throw has to handle whatever it is.

Reply:
Ok, regarding `SBThrow` (i.e. question 1), I think this is fine. It's an implementation detail whether to wrap the exception up, so as long as it makes sense and respects the semantics, it shouldn't matter. 

However, regarding the message function (i.e. question 2), I think moving `compileExceptionType` out of the compiler and doing the type-based dispatch only in `anyExceptionMessage` makes more sense than precomputing it here and above and then having to carry it all over the place.

You will presumably sometimes get exception values at run time (e.g. when communicating with the ledger ... or when running daml script and getting the result of an unhandled exception), and there you're only transmitting the type and fields. So computing the message function only when needed makes more sense to me.

Reply:
That said ... I think it's fine to leave it for now and maybe just keep in mind for future that it might be better to compute the message function only inside the `anyExceptionMessage` primitive.

Reply:
Ok. I didn't change anything for the moment.